Zend/zend_alloc.c

```diff
@@ -1904,6 +1967,27 @@ static void zend_mm_free_huge(zend_mm_heap *heap, void *ptr ZEND_FILE_LINE_DC ZE
 /* Initialization */
 /******************/
 
+static zend_result zend_mm_refresh_key(zend_mm_heap *heap)
+{
+	php_random_result result = php_random_algo_xoshiro256starstar.generate(&heap->random_state);
+	ZEND_ASSERT(result.size == sizeof(uint64_t));
+	heap->key = (uintptr_t) result.result;
+	return SUCCESS;
+}
+
+static zend_result zend_mm_init_key(zend_mm_heap *heap)
+{
+	uint64_t seed[4];
+	if (php_random_bytes(&seed, sizeof(seed), false) != SUCCESS) {
+		return FAILURE;
+	}
```

Comment:
A CSPRNG isn't strictly needed, a non-broken PRNG would be enough in this case.

Reply:
> A CSPRNG isn't strictly needed, a non-broken PRNG would be enough in this case.

To clarify: The CSPRNG is just used to seed the xoshiro256** instance, which is then used to generate the keys.

---------

My remark about the CSPRNG was a more general one: Should the CSPRNG be a hard requirement for PHP? It currently is not, e.g. here: https://github.com/php/php-src/blob/1cf4cc3894bfcecfc36b5681204d101992d76e5f/ext/session/session.c#L2874-L2881 or here: https://github.com/php/php-src/blob/1cf4cc3894bfcecfc36b5681204d101992d76e5f/ext/standard/uniqid.c#L77-L79

Reply:
> This will make the CSPRNG a hard dependency of ZendMM. As of now, a working CSPRNG is optional and various bits and pieces contain fallback logic by means of `php_random_generate_fallback_seed()`.

Thank you, I wasn't aware of that. I've added a fallback.